File: backend/app/schemas/mock_interview.py
"""Schemas for mock interview session and streaming APIs."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field, model_validator

from app.schemas.interview import Category, InterviewType
from app.schemas.jd import JDData
from app.schemas.resume import ResumeData
from app.schemas.runtime_config import RuntimeConfig

StreamMode = Literal["start", "reply"]
MessageRole = Literal["assistant", "user"]
MockInterviewCreateStage = Literal["retrieving_evidence", "generating_plan", "session_created"]


OPENING_ROUND_KEYWORDS = ("开场", "自我介绍", "背景", "介绍", "opening", "warm")
PROJECT_ROUND_KEYWORDS = ("项目", "项目概述", "经历", "经验", "overview", "experience")
CODING_ROUND_KEYWORDS = ("代码", "编码", "算法", "编程", "leetcode", "coding")


def _contains_keyword(text: str, keywords: tuple[str, ...]) -> bool:
    normalized = text.lower()
    return any(keyword.lower() in normalized for keyword in keywords)


class MockInterviewRound(BaseModel):
    """A single interview round with topic and description."""

    round: int = Field(ge=1, description="轮次编号")
    topic: str = Field(min_length=1, max_length=100, description="轮次主题")
    description: str = Field(min_length=1, max_length=300, description="轮次描述")


class MockInterviewPlan(BaseModel):
    """Canonical round/topic-driven interview plan."""

    plan: list[MockInterviewRound] = Field(min_length=3, max_length=12)
    total_rounds: int = Field(ge=3, le=12, description="总轮次数")
    estimated_duration: str = Field(min_length=1, max_length=40, description="预计时长")
    leetcode_problem: str = Field(min_length=1, max_length=200, description="代码题题目")

    @model_validator(mode="after")
    def validate_plan(self) -> "MockInterviewPlan":
        if len(self.plan) != self.total_rounds:
            raise ValueError("plan length must match total_rounds")

        expected_rounds = list(range(1, self.total_rounds + 1))
        actual_rounds = [item.round for item in self.plan]
        if actual_rounds != expected_rounds:
            raise ValueError("round numbers must be continuous from 1 to total_rounds")

        first_round_text = f"{self.plan[0].topic} {self.plan[0].description}"
        second_round_text = f"{self.plan[1].topic} {self.plan[1].description}"
        last_round_text = f"{self.plan[-1].topic} {self.plan[-1].description}"

        if not _contains_keyword(first_round_text, OPENING_ROUND_KEYWORDS):
            raise ValueError("first round must be an opening round")
        if not _contains_keyword(second_round_text, PROJECT_ROUND_KEYWORDS):
            raise ValueError("second round must be a project overview round")
        if not _contains_keyword(last_round_text, CODING_ROUND_KEYWORDS):
            raise ValueError("last round must be a coding round")

        return self


class MockInterviewRetrievalFilters(BaseModel):
    """Metadata filters applied to interview retrieval."""

    category: Category | None = None
    interviewType: InterviewType | None = None
    company: str | None = None


class MockInterviewRetrievalItem(BaseModel):
    """A single interview document used as planning evidence."""

    interviewId: int
    source: str = ""
    sourceId: str = ""
    title: str
    company: str | None = None
    category: Category
    interviewType: InterviewType | None = None
    stage: str | None = None
    publishTime: str = ""
    snippet: str = ""
    score: float = 0.0
    reason: str = ""


class MockInterviewRetrievalResult(BaseModel):
    """Interview evidence used during plan generation."""

    queryText: str = ""
    appliedFilters: MockInterviewRetrievalFilters = Field(default_factory=MockInterviewRetrievalFilters)
    items: list[MockInterviewRetrievalItem] = Field(default_factory=list)


class MockInterviewDeveloperContext(BaseModel):
    """Shared developer-facing context for local trace export."""

    sessionMode: Literal["frontend_local_only"] = "frontend_local_only"
    privacyMode: Literal["frontend_local_export_only"] = "frontend_local_export_only"
    ragEnabled: bool = True
    transcriptPersistence: Literal["frontend_local_only"] = "frontend_local_only"
    tracePersistence: Literal["frontend_local_only"] = "frontend_local_only"


class MockInterviewRetrievalTracePayload(BaseModel):
    """Developer trace payload for retrieval stage."""

    queryText: str = ""
    filterChain: list[MockInterviewRetrievalFilters] = Field(default_factory=list)
    appliedFilters: MockInterviewRetrievalFilters = Field(default_factory=MockInterviewRetrievalFilters)
    candidateTopk: int | None = None
    topk: int | None = None
    denseWeight: float | None = None
    sparseWeight: float | None = None
    ragEnabled: bool
    resultItems: list[MockInterviewRetrievalItem] = Field(default_factory=list)
    elapsedMs: int = Field(ge=0)


class MockInterviewPlanTracePayload(BaseModel):
    """Developer trace payload for plan generation stage."""

    promptKey: Literal["plan"] = "plan"
    jdDataIncluded: bool = False
    resumeProjectCount: int = Field(ge=0)
    retrievalItemCount: int = Field(ge=0)
    retrievalQueryText: str = ""
    outputPlan: MockInterviewPlan
    fallbackUsed: bool = False
    elapsedMs: int = Field(ge=0)


class MockInterviewReflectionTracePayload(BaseModel):
    """Developer trace payload for reflection stage."""

    promptKey: Literal["reflection"] = "reflection"
    candidateAnswer: str = Field(min_length=1)
    currentRoundHistory: str = ""
    questionCount: int = Field(ge=0)
    output: ReflectionResult
    fallbackUsed: bool = False
    elapsedMs: int = Field(ge=0)


class MockInterviewInterviewerTracePayload(BaseModel):
    """Developer trace payload for interviewer generation stage."""

    promptKey: Literal["interviewer"] = "interviewer"
    round: int = Field(ge=1)
    topic: str = ""
    suggestedFollowUp: str = ""
    closeInterview: bool = False
    recentConversation: list[dict[str, Any]] = Field(default_factory=list)
    finalMessage: str = ""
    elapsedMs: int = Field(ge=0)


class MockInterviewDeveloperTraceEvent(BaseModel):
    """Structured developer trace event streamed to the frontend."""

    type: Literal["retrieval", "plan_generation", "reflection", "interviewer_generation"]
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    payload: (
        MockInterviewRetrievalTracePayload
        | MockInterviewPlanTracePayload
        | MockInterviewReflectionTracePayload
        | MockInterviewInterviewerTracePayload
    )


class MockInterviewSessionLimits(BaseModel):
    """Runtime limits exposed to frontend."""

    durationMinutes: int = 60
    softInputChars: int = 1200
    maxInputChars: int = 1500
    contextWindowMessages: int = 8
    sessionTtlMinutes: int = 90


class MockInterviewMessage(BaseModel):
    """Stored transcript message."""

    id: str
    role: MessageRole
    content: str
    createdAt: datetime = Field(default_factory=datetime.utcnow)


class ReflectionResult(BaseModel):
    """Reflection evaluation result after each candidate answer."""

    depth_score: int = Field(ge=1, le=5, description="深度分数 (1-5)")
    authenticity_score: int = Field(ge=1, le=5, description="真实性分数 (1-5)")
    completeness_score: int = Field(ge=1, le=5, description="完整性分数 (1-5)")
    logic_score: int = Field(ge=1, le=5, description="逻辑性分数 (1-5)")
    overall_assessment: str = Field(min_length=10, max_length=300, description="整体评价")
    should_continue: bool = Field(description="是否继续当前轮次")
    suggested_follow_up: str = Field(default="", max_length=200, description="建议的追问方向")
    reason: str = Field(min_length=10, max_length=200, description="决策理由")

    @model_validator(mode="after")
    def validate_follow_up(self) -> "ReflectionResult":
        if self.should_continue and not self.suggested_follow_up.strip():
            raise ValueError("suggested_follow_up is required when should_continue is true")
        return self


class MockInterviewState(BaseModel):
    """Canonical runtime state for mock interview streaming."""

    currentRound: int = Field(default=1, ge=1)
    questionsPerRound: dict = Field(default_factory=dict)
    assistantQuestionCount: int = Field(default=0, ge=0)
    turnCount: int = Field(default=0, ge=0)
    reflectionHistory: list[ReflectionResult] = Field(default_factory=list)
    closed: bool = False

    @model_validator(mode="after")
    def normalize_questions_per_round(self) -> "MockInterviewState":
        normalized: dict[str, int] = {}
        for key, value in self.questionsPerRound.items():
            round_number = int(str(key))
            if round_number < 1:
                raise ValueError("questionsPerRound keys must be positive integers")
            if value < 0:
                raise ValueError("questionsPerRound values must be non-negative")
            normalized[str(round_number)] = value

        if str(self.currentRound) not in normalized:
            normalized[str(self.currentRound)] = 0

        self.questionsPerRound = normalized
        return self


class MockInterviewSessionCreateRequest(BaseModel):
    """Create a new mock interview session."""

    interviewType: InterviewType
    category: Category
    jdText: str = ""
    jdData: JDData | None = None
    resumeData: ResumeData
    runtimeConfig: RuntimeConfig | None = None

    @model_validator(mode="after")
    def validate_required_jd(self) -> "MockInterviewSessionCreateRequest":
        if not self.jdText.strip():
            raise ValueError("jdText is required for mock interview sessions")
        if self.jdData is None:
            raise ValueError("jdData is required for mock interview sessions")
        return self


class MockInterviewSessionCreateResponse(BaseModel):
    """Metadata returned after a session is created."""

    sessionId: str
    interviewType: InterviewType
    category: Category
    status: Literal["ready"] = "ready"
    limits: MockInterviewSessionLimits
    interviewPlan: MockInterviewPlan
    interviewState: MockInterviewState
    jdData: JDData | None = None
    retrieval: MockInterviewRetrievalResult = Field(default_factory=MockInterviewRetrievalResult)
    resumeFingerprint: str
    expiresAt: datetime
    developerContext: MockInterviewDeveloperContext | None = None


class MockInterviewCreateProgressEvent(BaseModel):
    """Progress payload for session creation stream."""

    stage: MockInterviewCreateStage
    message: str = Field(min_length=1, max_length=60)


class MockInterviewAnswerAnalysisStartedEvent(BaseModel):
    """Reply-phase status payload emitted before reflection begins."""

    stage: Literal["analyzing_answer"] = "analyzing_answer"
    message: str = Field(default="正在分析你的回答", min_length=1, max_length=60)


class MockInterviewStreamRequest(BaseModel):
    """Request body for streaming the next interviewer turn."""

    mode: StreamMode
    message: Optional[str] = None
    interviewType: InterviewType
    category: Category
    jdText: str = ""
    jdData: JDData | None = None
    resumeSnapshot: ResumeData
    runtimeConfig: RuntimeConfig | None = None
    retrieval: MockInterviewRetrievalResult = Field(default_factory=MockInterviewRetrievalResult)
    interviewPlan: MockInterviewPlan
    interviewState: MockInterviewState
    messages: list[MockInterviewMessage] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_request(self) -> "MockInterviewStreamRequest":
        if self.mode == "reply" and not (self.message or "").strip():
            raise ValueError("message is required when mode is 'reply'")
        if self.interviewState.currentRound > self.interviewPlan.total_rounds:
            raise ValueError("interviewState.currentRound cannot exceed interviewPlan.total_rounds")
        return self


class MockInterviewSessionSnapshot(BaseModel):
    """Frontend-local snapshot used as the single source for review generation."""

    snapshotVersion: int = Field(default=3, ge=1)
    sessionId: str
    interviewType: InterviewType
    category: Category
    status: Literal["ready", "streaming", "completed", "expired"]
    limits: MockInterviewSessionLimits
    jdText: str = ""
    jdData: JDData | None = None
    resumeSnapshot: ResumeData
    retrieval: MockInterviewRetrievalResult = Field(default_factory=MockInterviewRetrievalResult)
    interviewPlan: MockInterviewPlan
    interviewState: MockInterviewState
    messages: list[MockInterviewMessage] = Field(default_factory=list)
    developerContext: MockInterviewDeveloperContext | None = None
    developerTrace: list[MockInterviewDeveloperTraceEvent] = Field(default_factory=list)
    runtimeConfig: RuntimeConfig | None = None
    resumeFingerprint: str = ""
    createdAt: datetime
    lastActiveAt: datetime
    expiresAt: datetime

File: frontend/src/types/mockInterview.ts
export type MockInterviewStatus = "idle" | "creating" | "streaming" | "ready" | "completed" | "error";
export type MockInterviewRole = "assistant" | "user";
export type MockInterviewStreamMode = "start" | "reply";
export type MockInterviewCreatingStep = "idle" | "retrieving_evidence" | "generating_plan" | "starting_interview";
export type PendingAssistantPhase = "idle" | "analyzing_answer";

import type { Category, InterviewType } from "./interview";
import type { ResumeData } from "./resume";
import type { JDData, RuntimeConfig } from "@/lib/api";

export interface MockInterviewMessage {
  id: string;
  role: MockInterviewRole;
  content: string;
}

export interface MockInterviewLimits {
  durationMinutes: number;
  softInputChars: number;
  maxInputChars: number;
  contextWindowMessages: number;
  sessionTtlMinutes: number;
}

export interface MockInterviewRetrievalFilters {
  category: Category | null;
  interviewType: InterviewType | null;
  company: string | null;
}

export interface MockInterviewRetrievalItem {
  interviewId: number;
  title: string;
  company: string | null;
  category: Category;
  interviewType: InterviewType | null;
  stage: string | null;
  publishTime: string;
  snippet: string;
  score: number;
  reason: string;
}

export interface MockInterviewRetrievalResult {
  queryText: string;
  appliedFilters: MockInterviewRetrievalFilters;
  items: MockInterviewRetrievalItem[];
}

export interface ReflectionResult {
  depth_score: number;
  authenticity_score: number;
  completeness_score: number;
  logic_score: number;
  overall_assessment: string;
  should_continue: boolean;
  suggested_follow_up: string;
  reason: string;
}

export interface MockInterviewRound {
  round: number;
  topic: string;
  description: string;
}

export interface MockInterviewPlan {
  plan: MockInterviewRound[];
  total_rounds: number;
  estimated_duration: string;
  leetcode_problem: string;
}

export interface MockInterviewState {
  currentRound: number;
  questionsPerRound: Record<string, number>;
  assistantQuestionCount: number;
  turnCount: number;
  reflectionHistory: ReflectionResult[];
  closed: boolean;
}

export interface RoundTransition {
  from_round: number;
  to_round: number;
  topic: string;
}

export interface MockInterviewDeveloperContext {
  sessionMode: "frontend_local_only";
  privacyMode: "frontend_local_export_only";
  ragEnabled: boolean;
  transcriptPersistence: "frontend_local_only";
  tracePersistence: "frontend_local_only";
}

export interface MockInterviewRetrievalTracePayload {
  queryText: string;
  filterChain: MockInterviewRetrievalFilters[];
  appliedFilters: MockInterviewRetrievalFilters;
  candidateTopk: number | null;
  topk: number | null;
  denseWeight: number | null;
  sparseWeight: number | null;
  ragEnabled: boolean;
  resultItems: MockInterviewRetrievalItem[];
  elapsedMs: number;
}

export interface MockInterviewPlanTracePayload {
  promptKey: "plan";
  jdDataIncluded: boolean;
  resumeProjectCount: number;
  retrievalItemCount: number;
  retrievalQueryText: string;
  outputPlan: MockInterviewPlan;
  fallbackUsed: boolean;
  elapsedMs: number;
}

export interface MockInterviewReflectionTracePayload {
  promptKey: "reflection";
  candidateAnswer: string;
  currentRoundHistory: string;
  questionCount: number;
  output: ReflectionResult;
  fallbackUsed: boolean;
  elapsedMs: number;
}

export interface MockInterviewInterviewerTracePayload {
  promptKey: "interviewer";
  round: number;
  topic: string;
  suggestedFollowUp: string;
  closeInterview: boolean;
  recentConversation: Array<{ id: string; role: MockInterviewRole; content: string }>;
  finalMessage: string;
  elapsedMs: number;
}

export type MockInterviewDeveloperTraceEvent =
  | {
      type: "retrieval";
      createdAt: string;
      payload: MockInterviewRetrievalTracePayload;
    }
  | {
      type: "plan_generation";
      createdAt: string;
      payload: MockInterviewPlanTracePayload;
    }
  | {
      type: "reflection";
      createdAt: string;
      payload: MockInterviewReflectionTracePayload;
    }
  | {
      type: "interviewer_generation";
      createdAt: string;
      payload: MockInterviewInterviewerTracePayload;
    };

export interface MockInterviewSessionResponse {
  sessionId: string;
  interviewType: InterviewType;
  category: Category;
  status: "ready";
  limits: MockInterviewLimits;
  interviewPlan: MockInterviewPlan;
  interviewState: MockInterviewState;
  jdData: JDData | null;
  retrieval: MockInterviewRetrievalResult;
  resumeFingerprint: string;
  expiresAt: string;
  developerContext?: MockInterviewDeveloperContext | null;
}

export interface RuntimeAwarePayload {
  runtimeConfig?: RuntimeConfig | null;
}

export interface MockInterviewSessionSnapshot {
  sessionId: string;
  interviewType: InterviewType;
  category: Category;
  status: "ready" | "streaming" | "completed" | "expired";
  limits: MockInterviewLimits;
  jdText: string;
  jdData: JDData | null;
  resumeSnapshot: ResumeData;
  retrieval: MockInterviewRetrievalResult;
  interviewPlan: MockInterviewPlan;
  interviewState: MockInterviewState;
  messages: MockInterviewMessage[];
  developerContext: MockInterviewDeveloperContext | null;
  developerTrace: MockInterviewDeveloperTraceEvent[];
  pendingAssistantPhase?: PendingAssistantPhase;
  streamingMessageId?: string | null;
  runtimeConfig?: RuntimeConfig | null;
  resumeFingerprint: string;
  createdAt: string;
  lastActiveAt: string;
  expiresAt: string;
}

export interface MockInterviewPendingSession {
  pendingId: string;
  sessionId?: string | null;
  interviewType: InterviewType;
  category: Category;
  creatingStep: MockInterviewCreatingStep;
  startedAt: string;
  lastActiveAt: string;
}

export interface MockInterviewDeveloperReport {
  reportVersion: 1;
  exportedAt: string;
  privacy: {
    backendPersistence: "none";
    frontendPersistence: "local_storage_only";
  };
  session: {
    sessionId: string;
    status: MockInterviewSessionSnapshot["status"];
    interviewType: InterviewType;
    category: Category;
    createdAt: string;
    lastActiveAt: string;
    expiresAt: string;
    resumeFingerprint: string;
  };
  context: {
    developerContext: MockInterviewDeveloperContext | null;
    jdText: string;
    jdData: JDData | null;
    retrieval: MockInterviewRetrievalResult;
    interviewPlan: MockInterviewPlan;
    interviewState: MockInterviewState;
  };
  transcript: MockInterviewMessage[];
  developerTrace: MockInterviewDeveloperTraceEvent[];
  summary: {
    totalMessages: number;
    totalTraceEvents: number;
    currentRound: number;
    completed: boolean;
  };
}

export interface MockInterviewSessionCreateInput extends RuntimeAwarePayload {
  interviewType: InterviewType;
  category: Category;
  jdText: string;
  jdData?: JDData | null;
  resumeData: ResumeData;
}

export interface StreamCreateMockInterviewSessionInput {
  input: MockInterviewSessionCreateInput;
  signal?: AbortSignal;
  onProgress?: (payload: {
    stage: Exclude<MockInterviewCreatingStep, "idle" | "starting_interview">;
    message: string;
  }) => void | Promise<void>;
  onDeveloperTrace?: (payload: MockInterviewDeveloperTraceEvent) => void | Promise<void>;
  onSessionCreated?: (payload: MockInterviewSessionResponse) => void | Promise<void>;
  onDone?: (payload: { sessionId: string; status: "ready" }) => void | Promise<void>;
}

export interface StreamMockInterviewReplyInput extends RuntimeAwarePayload {
  sessionId: string;
  mode: MockInterviewStreamMode;
  interviewType: InterviewType;
  category: Category;
  jdText: string;
  jdData: JDData | null;
  resumeSnapshot: ResumeData;
  retrieval: MockInterviewRetrievalResult;
  interviewPlan: MockInterviewPlan;
  interviewState: MockInterviewState;
  messages: MockInterviewMessage[];
  message?: string;
  signal?: AbortSignal;
  onUserMessage?: (message: MockInterviewMessage) => void;
  onAnswerAnalysisStarted?: (payload: { stage: "analyzing_answer"; message: string }) => void;
  onMessageStart?: (payload: { messageId: string; role: MockInterviewRole }) => void;
  onMessageDelta?: (payload: { messageId: string; delta: string }) => void;
  onMessageEnd?: (payload: {
    messageId: string;
    content: string;
    interviewState: MockInterviewState;
    elapsedMs: number;
  }) => void;
  onDone?: (payload: {
    sessionId: string;
    status: "ready" | "completed";
    interviewState: MockInterviewState;
  }) => void;
  onReflection?: (reflection: ReflectionResult) => void;
  onDeveloperTrace?: (trace: MockInterviewDeveloperTraceEvent) => void;
  onRoundTransition?: (transition: RoundTransition) => void;
}
